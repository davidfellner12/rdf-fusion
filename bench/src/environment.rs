use crate::BenchmarkingOptions;
use crate::benchmarks::BenchmarkName;
use crate::prepare::{PrepRequirement, prepare_run_closure, prepare_run_command};
use crate::prepare::{ensure_file_download, prepare_file_download};
use anyhow::bail;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

/// Represents a context used to execute benchmarks.
pub struct RdfFusionBenchContext {
    /// General options for the benchmarks.
    options: BenchmarkingOptions,
    /// The path to the data dir.
    data_dir: Mutex<PathBuf>,
    /// The path to the current directory. This might be different from the `data_dir` if a
    /// benchmark is running.
    results_dir: Mutex<PathBuf>,
}

impl RdfFusionBenchContext {
    /// Creates a new [RdfFusionBenchContext].
    pub fn new(
        options: BenchmarkingOptions,
        data_dir: PathBuf,
        results_dir: PathBuf,
    ) -> Self {
        Self {
            options,
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(results_dir),
        }
    }

    /// Returns the [BenchmarkingOptions] for this context.
    pub fn options(&self) -> &BenchmarkingOptions {
        &self.options
    }

    /// Resolves a relative path `file` against the data directory.
    pub fn join_data_dir(&self, file: &Path) -> anyhow::Result<PathBuf> {
        if !file.is_relative() {
            bail!("Only relative paths can be resolved.")
        }

        Ok(self.data_dir.lock().unwrap().join(file))
    }

    /// Creates a new folder in the results directory and uses it until [Self::pop_dir] is
    /// called.
    ///
    /// This can be used to create folder hierarchies to separate the results of different
    /// benchmarks.
    #[allow(clippy::create_dir)]
    pub fn push_dir(&self, dir: &str) -> anyhow::Result<()> {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.push(dir);
        results_dir.push(dir);

        Ok(())
    }

    /// Pops the last directory from the stack.
    pub fn pop_dir(&self) {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.pop();
        results_dir.pop();
    }

    /// Creates a new bencher and modifies the context for this benchmark.
    pub fn create_benchmark_context(
        &self,
        benchmark_name: BenchmarkName,
    ) -> anyhow::Result<BenchmarkContext<'_>> {
        self.push_dir(&benchmark_name.dir_name())?;
        Ok(BenchmarkContext {
            context: self,
            benchmark_name,
        })
    }
}

/// A benchmarker that can be used to execute benchmarks.
///
/// It holds a reference to the current context to store its results.
pub struct BenchmarkContext<'ctx> {
    /// Reference to the benchmarking context.
    context: &'ctx RdfFusionBenchContext,
    /// Name of the benchmark that is being executed.
    benchmark_name: BenchmarkName,
}

impl<'ctx> BenchmarkContext<'ctx> {
    /// Returns a reference to the benchmarking context.
    pub fn parent(&self) -> &RdfFusionBenchContext {
        self.context
    }

    /// Returns the name of the benchmark that is being executed.
    pub fn benchmark_name(&self) -> BenchmarkName {
        self.benchmark_name
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn data_dir(&self) -> PathBuf {
        self.context.data_dir.lock().unwrap().clone()
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn results_dir(&self) -> PathBuf {
        self.context.results_dir.lock().unwrap().clone()
    }

    /// Prepares the context such that `requirement` is fulfilled.
    pub async fn prepare_requirement(
        &self,
        requirement: PrepRequirement,
    ) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload {
                url,
                file_name,
                action,
            } => prepare_file_download(self, url, file_name, action).await,
            PrepRequirement::RunClosure { execute, .. } => {
                prepare_run_closure(self, &execute)
            }
            PrepRequirement::RunCommand {
                workdir,
                program,
                args,
                ..
            } => {
                let workdir = self.context.join_data_dir(&workdir)?;
                prepare_run_command(&workdir, &program, &args)
            }
        }
    }

    /// Ensures that the `requirement` is fulfilled in this context.
    pub fn ensure_requirement(&self, requirement: PrepRequirement) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload { file_name, .. } => {
                ensure_file_download(self, file_name.as_path())
            }
            PrepRequirement::RunClosure {
                check_requirement, ..
            }
            | PrepRequirement::RunCommand {
                check_requirement, ..
            } => check_requirement(self),
        }
    }
}

/// Pops the results directory from the context when the bencher is dropped.
impl Drop for BenchmarkContext<'_> {
    fn drop(&mut self) {
        self.context.pop_dir();
    }
}
