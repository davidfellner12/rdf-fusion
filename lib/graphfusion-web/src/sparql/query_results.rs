use axum::response::{IntoResponse, Response};
use graphfusion::QueryResults;

/// Wraps a [QueryResults] that can be converted into a [Response].
pub struct QueryResultsResponse(QueryResults);

impl From<QueryResults> for QueryResultsResponse {
    fn from(value: QueryResults) -> Self {
        QueryResultsResponse(value)
    }
}

impl IntoResponse for QueryResultsResponse {
    fn into_response(self) -> Response {
        todo!()
    }
}
