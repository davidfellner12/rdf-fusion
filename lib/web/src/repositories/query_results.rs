use axum::body::{Body, Bytes};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use futures::{Stream, StreamExt};
use rdf_fusion::io::{RdfFormat, RdfSerializer};
use rdf_fusion::results::{QueryResultsFormat, QueryResultsSerializer};
use rdf_fusion::{QuerySolutionStream, QueryTripleStream};
use serde::de::StdError;
use std::pin::Pin;
use std::task::{ready, Context, Poll};

/// Wraps query results that can be converted into a [Response].
#[allow(unused)]
pub enum QueryResultsResponse {
    Solutions(QuerySolutionStream, QueryResultsFormat),
    Boolean(bool, QueryResultsFormat),
    Graph(QueryTripleStream, RdfFormat),
}

impl IntoResponse for QueryResultsResponse {
    fn into_response(self) -> Response {
        match self {
            QueryResultsResponse::Solutions(solutions, format) => {
                let solutions = QuerySolutionStreamResponseBody(solutions, format);
                let body = Body::from_stream(solutions);
                Response::builder()
                    .status(StatusCode::OK)
                    .body(body)
                    .unwrap()
            }
            QueryResultsResponse::Boolean(value, format) => {
                let mut buffer = Vec::new();
                let serializer = QueryResultsSerializer::from_format(format);
                serializer
                    .serialize_boolean_to_writer(&mut buffer, value)
                    .unwrap();

                Response::builder()
                    .header("Content-Type", "application/sparql-results+json")
                    .status(StatusCode::OK)
                    .body(Body::from(buffer))
                    .unwrap()
            }
            QueryResultsResponse::Graph(triples, format) => {
                let solutions = QueryTripleStreamResponseBody(triples, format);
                let body = Body::from_stream(solutions);
                Response::builder()
                    .status(StatusCode::OK)
                    .body(body)
                    .unwrap()
            }
        }
    }
}

struct QuerySolutionStreamResponseBody(QuerySolutionStream, QueryResultsFormat);

impl Stream for QuerySolutionStreamResponseBody {
    type Item = Result<Bytes, Box<dyn StdError + Send + Sync + 'static>>;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let mut buffer = Vec::new();
        let mut serializer = QueryResultsSerializer::from_format(self.1)
            .serialize_solutions_to_writer(&mut buffer, self.0.variables().to_vec())
            .unwrap();

        let mut count = 0;
        while let Some(solution) = ready!(self.0.poll_next_unpin(cx)) {
            serializer.serialize(solution.unwrap().into_iter()).unwrap();
            count += 1;

            if count == 1024 {
                break;
            }
        }

        if count == 0 {
            return Poll::Ready(None);
        }

        Poll::Ready(Some(Ok(Bytes::from(buffer))))
    }
}

struct QueryTripleStreamResponseBody(QueryTripleStream, RdfFormat);

impl Stream for QueryTripleStreamResponseBody {
    type Item = Result<Bytes, Box<dyn StdError + Send + Sync + 'static>>;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let mut buffer = Vec::new();
        let serializer = RdfSerializer::from_format(self.1);
        let mut serializer = serializer.for_writer(&mut buffer);

        while let Some(triple) = ready!(self.0.poll_next_unpin(cx)) {
            serializer
                .serialize_triple(triple.unwrap().as_ref())
                .unwrap();
        }

        serializer.finish().unwrap();

        Poll::Ready(Some(Ok(Bytes::from(buffer))))
    }
}
