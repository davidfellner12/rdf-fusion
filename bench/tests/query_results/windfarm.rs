//! This file contains tests for an adapted version of the Wind Farm benchmark. The adaption have
//! been made such that the queries produce stable results. The results of the queries has been
//! compared to GraphDB 11 to validate their correctness. See [crate::bsbm] for a detailed
//! description.

use crate::query_results::run_select_query;
use insta::assert_snapshot;
use rdf_fusion_bench::benchmarks::Benchmark;
use rdf_fusion_bench::benchmarks::windfarm::{NumTurbines, WindFarmBenchmark};
use rdf_fusion_bench::environment::RdfFusionBenchContext;
use std::path::PathBuf;

#[tokio::test]
pub async fn wind_farm_4_test_results() {
    let benchmarking_context =
        RdfFusionBenchContext::new_for_criterion(PathBuf::from("./data"), 1);
    let benchmark = WindFarmBenchmark::new(NumTurbines::N4);
    let benchmark_name = benchmark.name();
    let ctx = benchmarking_context
        .create_benchmark_context(benchmark_name)
        .unwrap();

    let store = benchmark.prepare_store(&ctx).await.unwrap();

    assert_snapshot!(
        "Production Q1",
        run_select_query(
            &store,
            include_str!("./queries/wind-farm-production-query1.sparql")
        )
        .await
    );
    assert_snapshot!(
        "Production Q2",
        run_select_query(
            &store,
            include_str!("./queries/wind-farm-production-query2.sparql")
        )
        .await
    );
    assert_snapshot!(
        "Production Q3",
        run_select_query(
            &store,
            include_str!("./queries/wind-farm-production-query3.sparql")
        )
        .await
    );
    assert_snapshot!(
        "Production Q4",
        run_select_query(
            &store,
            include_str!("./queries/wind-farm-production-query4.sparql")
        )
        .await
    );

    assert_snapshot!(
        "Grouped Production Q1",
        run_select_query(
            &store,
            include_str!("./queries/wind-farm-grouped-production-query1.sparql")
        )
        .await
    );
    assert_snapshot!(
        "Grouped Production Q2",
        run_select_query(
            &store,
            include_str!("./queries/wind-farm-grouped-production-query2.sparql")
        )
        .await
    );
    assert_snapshot!(
        "Grouped Production Q3",
        run_select_query(
            &store,
            include_str!("./queries/wind-farm-grouped-production-query3.sparql")
        )
        .await
    );
    assert_snapshot!(
        "Grouped Production Q4",
        run_select_query(
            &store,
            include_str!("./queries/wind-farm-grouped-production-query4.sparql")
        )
        .await
    );
}
