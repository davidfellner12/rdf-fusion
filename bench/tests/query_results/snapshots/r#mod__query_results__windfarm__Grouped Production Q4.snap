---
source: bench/tests/query_results/windfarm.rs
expression: "run_select_query(&store,\ninclude_str!(\"./queries/wind-farm-grouped-production-query4.sparql\")).await"
---
{
  "head": {
    "vars": [
      "site_label",
      "wtur_label",
      "year",
      "month",
      "day",
      "hour",
      "minute_10",
      "avg_val"
    ]
  },
  "results": {
    "bindings": [
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7796289.200858504"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "40"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7795129.20661896"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "50"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7793161.64541853"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "9"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "0"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7782867.750758753"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "9"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "10"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7789637.599417424"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "9"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "20"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7788157.5312966965"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "9"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "30"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7787551.409267022"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "9"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "40"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7780092.301139752"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "9"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "50"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7777626.313333333"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "10"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "0"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      },
      {
        "avg_val": {
          "datatype": "http://www.w3.org/2001/XMLSchema#double",
          "type": "literal",
          "value": "7774602.599276077"
        },
        "day": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "30"
        },
        "hour": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "10"
        },
        "minute_10": {
          "datatype": "http://www.w3.org/2001/XMLSchema#decimal",
          "type": "literal",
          "value": "10"
        },
        "month": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "8"
        },
        "site_label": {
          "type": "literal",
          "value": "Gale Valley"
        },
        "wtur_label": {
          "type": "literal",
          "value": "A1"
        },
        "year": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "2022"
        }
      }
    ]
  }
}
